oap-server/server-storage-plugin/storage-elasticsearch-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch/base/StorageEsInstaller.java

```diff
@@ -305,38 +305,40 @@ private Map getAnalyzerSetting4MergedIndex(Model model) throws StorageException
     }
 
     protected Mappings createMapping(Model model) {
-        Map<String, Object> properties = new HashMap<>();
-        Mappings.Source source = new Mappings.Source();
+        final var properties = new HashMap<String, Object>();
+        final var source = new Mappings.Source();
         for (ModelColumn columnDefine : model.getColumns()) {
+            final var elasticSearchExtension = columnDefine.getElasticSearchExtension();
             final String type = columnTypeEsMapping.transform(columnDefine.getType(), columnDefine.getGenericType(),
                 columnDefine.getLength(), columnDefine.isStorageOnly(),
-                columnDefine.getElasticSearchExtension());
-            String columnName = columnDefine.getColumnName().getName();
-            String legacyName = columnDefine.getElasticSearchExtension().getLegacyColumnName();
+                elasticSearchExtension);
+            var columnName = columnDefine.getColumnName().getName();
+            final var legacyName = elasticSearchExtension.getLegacyColumnName();
+            final var columnProperties = new HashMap<>();
             if (config.isLogicSharding() && !Strings.isNullOrEmpty(legacyName)) {
                 columnName = legacyName;
             }
-            if (columnDefine.getElasticSearchExtension().needMatchQuery()) {
+            if (elasticSearchExtension.needMatchQuery()) {
                 String matchCName = MatchCNameBuilder.INSTANCE.build(columnName);
 
-                Map<String, Object> originalColumn = new HashMap<>();
-                originalColumn.put("type", type);
-                originalColumn.put("copy_to", matchCName);
-                properties.put(columnName, originalColumn);
+                columnProperties.put("type", type);
+                columnProperties.put("copy_to", matchCName);
 
                 Map<String, Object> matchColumn = new HashMap<>();
                 matchColumn.put("type", "text");
-                matchColumn.put("analyzer", columnDefine.getElasticSearchExtension().getAnalyzer().getName());
+                matchColumn.put("analyzer", elasticSearchExtension.getAnalyzer().getName());
                 properties.put(matchCName, matchColumn);
             } else {
-                Map<String, Object> column = new HashMap<>();
-                column.put("type", type);
+                columnProperties.put("type", type);
                 // no index parameter is allowed for binary type, since ES 8.0
                 if (columnDefine.isStorageOnly() && !"binary".equals(type)) {
-                    column.put("index", false);
+                    columnProperties.put("index", false);
                 }
-                properties.put(columnName, column);
             }
+            if (!elasticSearchExtension.isDocValuesEnabled()) {
+                columnProperties.put("doc_values", false);
+            }
```

Comment:
I think we have OAL generated metrics' fields for service name, etc. Do we make sure what works?